src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,758 @@
+
+	logicalClockShmem = (LogicalClockShmemData *)
+						ShmemInitStruct("Logical Clock Shmem",
+										LogicalClockShmemSize(),
+										&alreadyInitialized);
+
+	if (!alreadyInitialized)
+	{
+		/* A zero value indicates that the clock is not adjusted yet */
+		logicalClockShmem->clusterClockValue = 0;
+		SpinLockInit(&logicalClockShmem->clockMutex);
+		logicalClockShmem->clockInitialized = CLOCKSTATE_UNINITIALIZED;
+	}
+
+	LWLockRelease(AddinShmemInitLock);
+
+	if (prev_shmem_startup_hook != NULL)
+	{
+		prev_shmem_startup_hook();
+	}
+}
+
+
+/*
+ * GetNextClusterClock implements the internal guts of the UDF citus_get_cluster_clock()
+ */
+ClusterClock
+GetNextClusterClockValue(void)
+{
+	uint64 wallClockValue;
+	uint64 epochValue = GetEpochTimeMs();
+
+	SpinLockAcquire(&logicalClockShmem->clockMutex);
+
+	/* Check if the clock is adjusted after the boot */
+	if (logicalClockShmem->clockInitialized == CLOCKSTATE_UNINITIALIZED)
+	{
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+		Assert(logicalClockShmem->clusterClockValue == 0);
+		InitClockAtFirstUse();
+
+		SpinLockAcquire(&logicalClockShmem->clockMutex);
+		if (logicalClockShmem->clockInitialized == CLOCKSTATE_INIT_INPROGRESS ||
+			logicalClockShmem->clockInitialized == CLOCKSTATE_UNINITIALIZED)
+		{
+			/* Either we lost the initialization-race or there was an exception */
+			SpinLockRelease(&logicalClockShmem->clockMutex);
+			ereport(ERROR, (errmsg("Clock is in the process of getting initialized, "
+								   "please retry")));
+		}
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+	}
+
+	Assert(logicalClockShmem->clockInitialized == CLOCKSTATE_INITIALIZED);
+
+	/* Set the epoch in (lc, c) format */
+	SET_CLOCK(wallClockValue, epochValue, 0);
+
+	uint64 nextClusterClockValue = Max(logicalClockShmem->clusterClockValue + 1,
+									   wallClockValue);
+
+	logicalClockShmem->clusterClockValue = nextClusterClockValue;
+
+	SpinLockRelease(&logicalClockShmem->clockMutex);
+
+	/* Return ClusterClock custom type */
+	ClusterClock clusterClock = INIT_CLUSTER_CLOCK(nextClusterClockValue);
+	return clusterClock;
+}
+
+
+/*
+ * IsClockAfter implements the internal guts of the UDF citus_is_clock_after()
+ */
+static bool
+IsClockAfter(ClusterClock clusterClock1, ClusterClock clusterClock2)
+{
+	ereport(DEBUG1, (errmsg(
+						 "clock1 @ LC:%lu, C:%u, "
+						 "clock2 @ LC:%lu, C:%u",
+						 clusterClock1.logical, clusterClock1.counter,
+						 clusterClock2.logical, clusterClock2.counter)));
+
+	if (clusterClock1.logical != clusterClock2.logical)
+	{
+		return (clusterClock1.logical > clusterClock2.logical);
+	}
+	else
+	{
+		return (clusterClock1.counter > clusterClock2.counter);
+	}
+}
+
+
+/*
+ * AdjustLocalClock Adjusts the local shared memory clock to the
+ * received value from the remote node.
+ */
+void
+AdjustLocalClock(ClusterClock remoteClock)
+{
+	uint64 remoteClusterClock = { 0 };
+
+	SET_CLOCK(remoteClusterClock, remoteClock.logical, remoteClock.counter);
```

Comment:
We still have two representations of the cluster clock.
 - `ClusterClock`
 - `uint64`
 
 What is the benefit of having two?

Reply:
I wanted to start like this, which will avoid lot of bit-operations and also code looks clean, PG community uses such syntax, but I haven't seen it in Citus code. I was skeptic if I can use that style and posted it in our internal channel, but was discouraged.

```
typedef struct ClusterClock
{
        uint64 logical : 44;     /* cluster clock logical timestamp at the commit */
        uint64 counter : 22;    /* cluster clock counter value at the commit */
} ClusterClock;

```

Reply:
I have been looking into that notation above a bit as well and came to the conclusion it is pretty hard to use. If you want to use it for efficient storage I think you will also need to pass the packed attribute.

Unfortunately this is where a big discrepancy comes between different endian machines. On x86_64 with my compiler on OSX it caused the counter to be put in the more significant bits of the 64bit space and the logical into the lower space. This is incoherent with the normal memory model C provides us with and caused me some headaches as well.

The internet was mostly down to, it is compiler dependant on how this will be stored, and you can't really rely on storing this on disk and read it back forever and be correct.

Reply:
Also, for storing the clock value elsewhere the 64 bit limit is still a sane limit.